### PR TITLE
Add Multi-GPU training, Griffin-Lim vocoder, safely restore checkpoints, pathlib, multiple hparams files

### PR DESCRIPTION
**Pull request by TheButlah**
Follows up on PR #109. 

Added the following features:
- Multi-GPU Training.
- Added Griffin-Lim as a baseline vocoder
- Restoring checkpoints now also accounts for optimizer state, making it safe to save and restore training sessions
- `hparams` module can be specified via argument to facilitate testing with different sets of hyperparameters
- Vocoder and Tacotron both use gradient clipping by default now, and can be configured in `hparams`

Refactored the following:
- Codebase now uses pathlib to be more platform independent and nice to use
- Paths module is now clearer and less ambiguous by using pathlib, no longer need to worry if the paths should end in `/` or not
- More things that should have been in PyTorch buffers are now in buffers
- Checkpoint saving and restoring completely changed, explicitly delegates the responsibility of training-related state to the training scripts, and model-related state to the models.
- Specifying training vs eval mode uses PyTorch conventions

Fixed the following bugs:
- Issue where saving spectrograms via matplotlib was using interactive backend, requiring X11 to be running.
- Missing imports
- Other bugs

~**Please vet this code! It's not yet thoroughly tested.**~

Another caveat to note: The notebooks are now very very outdated. I haven't touched them, and they appear to use their own code so none of this should affect them.

### CODE REVIEW

**Comment by TheButlah** [comment]
One caveat that I should note: For some reason, I don't always experience speedups when attempting to run on multiple GPUs. I think this might just be because my GPUS are so fast and the batch size is relatively small, so the overhead of the data parallel code is hiding any benefits one might see.

This isn't really an issue, just be sure to compare single-GPU vs multiple-GPU on your system to see which works better for you.

**Comment by fatchord** [comment]
This looks awesome @TheButlah - I'm still training models. It'll just take a bit of time.

**Comment by TheButlah** [comment]
@fatchord Glad to help!

**Comment by TheButlah** [comment]
I just completed a training session of WaveRNN (without Taco) and even after all 1 million steps, I still get bad results. It seems like every epoch it alternates between good results and bad results.

Here are some samples. Note that I have modified some hparams, which may be the cause of all of this.
[Results.zip](https://github.com/fatchord/WaveRNN/files/3487352/Results.zip)

**Comment by TheButlah** [comment]
Across the board whether its your parameters or mine, I see poor results. I would greatly appreciate any insights you have on why the results seem of low quality - I'm out of ideas. I've looked through the codebase but I'm not really seeing anything.

**Comment by TheButlah** [comment]
[Archive.zip](https://github.com/fatchord/WaveRNN/files/3488356/Archive.zip)
Here are some outputs of tacotron trained with your hparams using Griffin-Lim as a vocoder. They seem more reasonable but still low quality. Are these results in line with what your results were?

**Comment by TheButlah** [comment]
Update: I just added in Griffin-Lim as an option to use for the Tacotron vocoder. This allows testing WaveRNN and tacotron independently. I also just added backwards comaptibility support for loading the old models where the r value was a parameter under Tacotron instead of a buffer under tacotron.decoder.

As far as isolating the bug that is causing the poor training performance, I think it definately has to something to do with the gradients. I think the most likely culprit could be the gradient clipping I added in - perhaps this hinders the model's weights from traveling as far as they should be. Another alternative is the multi-GPU data parallel. Maybe I did something wrong there and some data is not getting synced between devices somehow.

Will let you know if I find anything. Right now I have only a bunch of different models to go off of, but no solid evidence.

**Comment by TheButlah** [comment]
Update: Added `tts_stop_threshold` hparam that controls the threshold at which tacotron stops generating. This value was originally hardcoded at -3.9, but I've made it -3.4 because that seemed to be better. Also made `--force_train` in tacotron actually force training.

As far as the convergence issues are concerned, **I've come to the conclusion that I don't think the code is bugged.** I think this is a hyperparameter issue. I have compared my retrained version with your hparams with your pretrained model. They are about equivalent at 180K steps, when `r` was equal to 2. However, training for the full schedule that you specified in `hparams.py` causes poor convergence. I suspect the reason is one or both of two possibilities:
1. Training for too many iterations causes overfitting
2. Setting `r` to 1 might need even more iterations to properly converge, or its just not a good choice. The original Tacotron paper used `r=2` after all.

I don't know which of those two is true. I'm curious to see what your results are when you retrain. Let me know :)

Note that all of this investigation has focused on Tacotron only, I have not vetted WaveRNN yet but I don't see why it wouldn't apply to it too.

**Comment by TheButlah** [comment]
It sounds about equivalent to yours, for Tacotron (again haven't vetted WaveRNN but it shares a lot of its code with Tacotron so its probably working too) which means its probably working. Here are some griffin lim samples generated with your original tacotron pretrained model (no retraining):
[Author Hparams Original.zip](https://github.com/fatchord/WaveRNN/files/3493502/Author.Hparams.Original.zip)

And here is the same hparams trained to the same number of iters, but retrained on my fork:
[Author Hparams Retrained (same training).zip](https://github.com/fatchord/WaveRNN/files/3493506/Author.Hparams.Retrained.same.training.zip)

As you can see, the quality significantly degrades when you continue to train on your hparams for the full 350 iters:
[Authors Hparams (longer training and lower r).zip](https://github.com/fatchord/WaveRNN/files/3493512/Authors.Hparams.longer.training.and.lower.r.zip)

**Comment by TheButlah** [comment]
@fatchord yep I just took a look at the results from continuing to train over the weekend, and the version where r is reduced to 1 fails to converge even at 467K iters, whereas the version where I keep r=2 continues to improve (at 296K iters so far).

I would suggest updating the learning schedule to change r=1 in the last session to r=2. This works dramatically better in my tests.

I will start training WaveRNN today or tomorrow, so I can confirm the correctness of the vocoder soon too

**Comment by TheButlah** [comment]
@fatchord I have retrained both WaveRNN and Tacotron and have confirmed that they work. Here are the results, when using the tacotron inference script with wavernn as the vocoder. WaveRNN has not yet finished training. I've included both Griffin-Lim and WaveRNN for the sake of comparison. Tacotron trained with r=2 in the last session, with a total number of iters of 376K. WaveRNN not yet finished training, with a total number of iters of 650K (not even 2/3 trained!).

These results should prove that this fork is tested and ready to merge in. Let me know if you have any questions or need help with my changes!
[Archive.zip](https://github.com/fatchord/WaveRNN/files/3502677/Archive.zip)

**Comment by TheButlah** [comment]
Update: finished training wavernn. Results good. Did notice that the quality seems to randomly deteriorate a little depending on which checkpoint I'm looking at. I think this is because the learning rate should be decreased later over the course of training. I trained to 1050K steps with the regular 1e-4 learning rate, and will train for another 200k steps with 1e-5 learning rate. The results at 1.1M steps already sound good, so this was probably a good call.

 It would probably make sense to use training sessions like tacotron uses, but I'm not going to add that for now because its easy enough to find a checkpoint that sounds good, and the deterioration in quality is pretty minimal.

Again, let me know if you have any questions about the patch, it is ready to be merged in whenever.

**Comment by fatchord** [comment]
@TheButlah Wow, that's amazing, thanks - can I get the weights to try out?

**Comment by fatchord** [comment]
Btw - sorry I haven't been more active on this - work has been full on the last couple of weeks.

**Comment by TheButlah** [comment]
Hi @fatchord! No problem, its your repo after all, I've already got everything working on my end so its not like you're holding me up :)

I cannot release my trained models unfortunately, but you should be able to just use the edits I made to the hparams on my fork to get the results I posted for Tacotron, and if you do the trick I mentioned about WaveRNN with the lower learning rate after training with the lr in hparams.py, you should be able to reproduce my results.

One way or another though, the fork should work 100% as it stands and is backwards compatible with your pretrained models. Feel free to merge it in as it is once you've tested it :)

**Comment by TheButlah** [comment]
@fatchord any updates on the status of the merge?

**Comment by fatchord** [comment]
@TheButlah Sorry about the delay - I only got around at looking at it last night. There's a couple of small things I'd like to change - I want to put the new checkpointing functions into utils as there's duplicate code in the training files. Other than that - just small things and it's good to merge. 

I'll create a new branch from your fork and get working on that tomorrow morning. Hopefully we can get this merged by the weekend.

**Comment by fatchord** [comment]
Ok I did a little bit there. Have a look and see if that makes sense. There might be a few more things - I'll have another look tomorrow morning.

**Comment by TheButlah** [comment]
Looks good. I don't have my development environment set up right now for this project, so I'll take your word for it that it works. I did make one change to actually use the real r value in quick_start.py.

**Comment by fatchord** [comment]
I think it's good to merge now. @TheButlah Thanks again for your awesome contribution.
